---
title: "R Notebook"
output: html_notebook
---

```{r}
knitr::opts_knit$set(root.dir = normalizePath("/Users/nataliefortunato/Documents/qbio_490_nataliefortunato/analysis_data"))
```

```{r}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.16")

if (!require("TCGAbiolinks", quietly = TRUE))
  BiocManager::install("TCGAbiolinks")

if (!require("maftools", quietly = TRUE))
  BiocManager::install("maftools")

library(BiocManager)

library(TCGAbiolinks)

library(maftools)
```

```{r}
clin_query <- GDCquery(project = "TCGA-BRCA", data.category = "Clinical", file.type = "xml")

#GDCdownload(clin_query)

clinic <- GDCprepare_clinic(clin_query, clinical.info = "patient")
```
```{r}
colnames(clinic)[ colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

maf_query <- GDCquery(
  project = "TCGA-BRCA",
  data.category = "Simple Nucleotide Variation",
  access = "open",
  data.type = "Masked Somatic Mutation",
  workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

GDCdownload(maf_query)

maf <- GDCprepare(maf_query)

maf_object <- read.maf(maf = maf, clinicalData = clinic, isTCGA = TRUE)
```
```{r}
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
#GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```




